Address selector on mobile

onmax opened the pull request:
## Current state

If user is using a mobile device and the account column is active, the buttons "receive" and "send" will use the last selected address as a recipient or as a sender. This is not very intuitive.

[See video](https://www.loom.com/share/d5dec9857bc54d538aeb7811a47df6d1)

## Changes in this PR

With this PR, if the user is using a mobile device and account column is active, when clicking "receive" and "send", user can select a NIM address or BTC

[See video](https://www.loom.com/share/a3d0fcc419fc429192979211c8412bcf)

## Changes in code

1. Added new composable `useActiveMobileColumn`
2. New way to close modal to support user stories that contain multiple modal one after the other
3. Added `AddressSelectorModal`
4. Refactor `/send` and `/receive` paths
5. Added back arrow to some screens if the user is coming from the address selector and updated some screens repositioning some icons

## Future work

- User in 'Choose a Sender' should not be able to select addresses with 0 assets
- In `BTCSendModal`, back arrow and info icon are shown very close to each other, it might be subject of change in the future

onmax committed:
Address selector has been added for receive operations:
- modified receive paths to: /receive, /receive/nim and /receive/btc
- new composable useActiveMobileColumn
- user can select in which address he will receive the assets if he is on the mobile and the address column is active

Co-authored-by: Daniel <github.danimoh@neverbox.com>

onmax committed:
Address selector has been added for send operations
- modified receive paths to: /send, /send/nim and /send/btc
- user can select to which address to send the assets if he/she is on the mobile and the address column is active

danimoh committed:
Mobile address selector: some clean up and fixes

onmax committed:
closing modal goes back continuously until no modal is open

onmax committed:
hide back arrow when is a requestUri for nim

onmax committed:
hide back arrow in btc request uri

sisou commented on src/components/MobileActionBar.vue:
```diff
@@ -38,14 +42,24 @@ export default defineComponent({
+                // redirect to the address selector
+                context.root.$router.push('/receive');
+            } else {
+                context.root.$router.push(nimOrBtc<string>('/receive/nim', '/receive/btc'));
```
The `<string>` specification of the generic type of the `nimOrBtc` function is actually not necessary, Typescript figures it out itself.

sisou commented on src/components/modals/AddressSelectorModal.vue:
```diff
@@ -0,0 +1,65 @@
+                {{ $t("Choose a Sender")}}
+            </template>
+            <template v-else-if="path === '/receive'">
+                {{ $t("Choose a Receiver")}}
```
Maybe "Choose a Recipient"? We can ask Elion also, he usually has an opinion on this :D

sisou commented on src/components/modals/AddressSelectorModal.vue:
```diff
@@ -0,0 +1,65 @@
+        <PageBody>
+            <AddressList
+                embedded
+                showBitcoin
```
Legacy accounts do not have Bitcoin, and iOS or Ledger accounts that have not activated it also don't have a BTC account. So this needs to be conditional, and only shown when BTC is available/enabled on this account.

sisou commented on src/components/MobileActionBar.vue:
```diff
@@ -38,14 +42,24 @@ export default defineComponent({
+            if (width.value <= 700 /* Full mobile breakpoint */ && activeMobileColumn.value !== ColumnType.ADDRESS) {
+                // redirect to the address selector
+                context.root.$router.push('/receive');
+            } else {
```
Accounts that only have one eligible address (read: only one regular address, contracts don't count) and don't have BTC available/enabled, we should skip the selection and go directly to /receive/nim.

sisou commented on src/components/MobileActionBar.vue:
```diff
@@ -38,14 +42,24 @@ export default defineComponent({
+            if (width.value <= 700 /* Full mobile breakpoint */ && activeMobileColumn.value !== ColumnType.ADDRESS) {
+                // redirect to the address selector
+                context.root.$router.push('/send');
+            } else {
```
If only one address exists on the account, and BTC is not available/enabled, skip the selection and go directly to /send/nim.

sisou commented on src/components/modals/BtcReceiveModal.vue:
```diff
@@ -317,6 +324,14 @@ export default defineComponent({
             window.removeEventListener('resize', updateAddressFontSizeScaleFactor);
         });
 
+        // User only can go back to the address selection if is mobile and the column shown is the account
```
```suggestion
        // User can only go back to the address selection if is mobile and the column shown is the account
```

sisou commented on src/components/modals/BtcReceiveModal.vue:
```diff
@@ -317,6 +324,14 @@ export default defineComponent({
         });
 
+        // User only can go back to the address selection if is mobile and the column shown is the account
+        const canUserGoBack = ref(width.value <= 700 && activeMobileColumn.value !== ColumnType.ADDRESS);
```
This way of doing it works for now, but may not hold up in the future when we might have new ways to reach this modal. 

This way right now is _implicit_, where the receive modal itself determines if it should show the back button or not, depending on where it's shown. It does not consider where the user came from.

I think a better way would be to make this choice _explicit_ to the function call that opens this route. This modal can take a prop that determines if the back-arrow should be shown, the prop can be passed as a `param` in the router.push call.

danimoh replied:
But the param would be visible in the url, which is not so nice.
It could also be stored in the history state when switching the modal.

sisou replied:
This is false. There is a `param` option for router methods, that pass as props, not as `query` params.
https://router.vuejs.org/guide/essentials/passing-props.html
https://router.vuejs.org/api/#route-object-properties

onmax replied:
I think there are only two ways of passing variables between pages in Vue and both of them use the URI as a channel :
1. `URL params`: As an example `/user/onmax` where `onmax` could be the variable, and then in the router you indicate Vue how you want to map the value
2. `Query param`: As an example `/user?username=onmax` where username is the key that the router will use to map the value

Therefore, what you are proposing is modifying the URI to something like `/receive/nim/canUserGoBack`, and @danimoh and me discarded that option. If I am not wrong, the only way to pass data between routes in Vue is using the URI using `URL params` or `query params`

sisou replied:
We found that using router params still works when used with specifying the route by its `name`, and if no placeholder is set in the route's path, the param does not show up there, but is still available on `$router.params`.

sisou commented on src/components/modals/BtcReceiveModal.vue:
```diff
@@ -362,12 +379,77 @@ export default defineComponent({
         font-weight: 600;
-        opacity: 0.6;
-        margin-top: 2rem;
+        text-align: initial;
```
Why is `text-align` needed? Removing it makes no visible difference for me.

danimoh replied:
To align the text to the left when it breaks into two lines.

sisou commented on src/components/modals/BtcSendModal.vue:
```diff
@@ -494,16 +498,7 @@ export default defineComponent({
-                        context.root.$router.back();
-                    }
-                }, SUCCESS_REDIRECT_DELAY);
+                sucessCloseTimeout = window.setTimeout(() => $modal.value!.forceClose(), SUCCESS_REDIRECT_DELAY);
```
Typo: `sucess` => `success`

sisou commented on src/components/modals/Modal.vue:
```diff
@@ -150,8 +176,15 @@ export default defineComponent({
         }
 
+        onMounted(() => {
+            setTimeout(() => {
+                enableModalTransition();
+            }, 100);
```
Why is the timeout required here?

danimoh replied:
At the time the modal instance gets created, the transition still needs to be disabled to avoid that showing this modal is animated.

sisou commented on src/composables/useActiveMobileColumn.ts:
```diff
@@ -0,0 +1,53 @@
+}, 0);
+
+export function useActiveMobileColumn() {
+    if (!activeMobileColumn) {
+        activeMobileColumn = ref(ColumnType.ACCOUNT);
+    }
```
Since the variable is initialized here when needed and also in the beforeResolve function above, wouldn't it make sense to directly initialize it to its Ref-value, instead of to null first?

danimoh replied:
The ref function can apparently not be called at that time yet.
cf https://github.com/nimiq/wallet/blob/master/src/composables/useWindowSize.ts#L5-L9

sisou replied:
Right. Can this comment be copied to here, too, so this is documented?

sisou commented on src/i18n/en.po:
```diff
@@ -151,15 +151,15 @@ msgstr ""
 msgstr ""
 
-#: src/components/modals/SendModal.vue:174
+#: src/components/modals/SendModal.vue:176
```
We usually try to exclude the language files from PRs, as they only create merge conflicts. They are rebuild when the app is deployed.

danimoh replied:
:see_no_evil:

sisou reviewed (changes requested):
Well done :+1:

One issue in regards to how the address-selection-modal handles legacy accounts and accounts without BTC, and one optional issue where I think the logic to show the back arrow can be handled better explicitly.

danimoh committed:
Mobile address selector: revert import order changes; update en.po

onmax committed:
changed to ::v-deep

onmax committed:
@sisou's suggestions in #72

- fixed typos and added comments
- closing the modal now uses a explicit function
- better management of legacy accounts and vesting contracts

Co-authored-by: sisou <hello@soerenschwert.de>

onmax commented:
@sisou all the change requests and suggestions have been made 🐒

sisou committed:
Use explicit back-arrow param in send modals, too

sisou committed:
Fix being unable to use the address selector for sending

danimoh commented on src/components/MobileActionBar.vue:
```diff
@@ -41,21 +41,36 @@ export default defineComponent({
             }
         }
 
+        const hasBitcoin = computed(() => Boolean(activeAccountInfo.value
+                && activeAccountInfo.value.btcAddresses && activeAccountInfo.value.btcAddresses.external.length));
```
Should be updated after https://github.com/nimiq/wallet/pull/74 is merged.

danimoh commented on src/components/MobileActionBar.vue:
```diff
@@ -41,21 +41,36 @@ export default defineComponent({
+
         function receive() {
-            if (width.value <= 700 /* Full mobile breakpoint */ && activeMobileColumn.value !== ColumnType.ADDRESS) {
+            if (width.value <= 700 /* Full mobile breakpoint */
```
Should be updated after https://github.com/nimiq/wallet/pull/74 is merged.

danimoh commented on src/components/MobileActionBar.vue:
```diff
@@ -41,21 +41,36 @@ export default defineComponent({
+
         function send() {
-            if (width.value <= 700 /* Full mobile breakpoint */ && activeMobileColumn.value !== ColumnType.ADDRESS) {
+            if (width.value <= 700 /* Full mobile breakpoint */
```
Should be updated after https://github.com/nimiq/wallet/pull/74 is merged.

danimoh commented on src/components/modals/AddressSelectorModal.vue:
```diff
@@ -32,17 +32,28 @@ export default defineComponent({
         }
 
+        const { activeAccountInfo } = useAccountStore();
+        const showBitcoin = computed(() => Boolean(activeAccountInfo.value
+            && activeAccountInfo.value.btcAddresses
+            && activeAccountInfo.value.btcAddresses.external.length));
```
Should be updated after https://github.com/nimiq/wallet/pull/74 is merged.

danimoh commented on src/components/modals/AddressSelectorModal.vue:
```diff
@@ -32,17 +32,28 @@ export default defineComponent({
+            context.root.$router.push({
+                name: `${name}-${activeCurrency.value}`,
+                params: {
+                    canUserGoBack: 'true',
```
Does `true` have to be a string?

Edit: I guess [this comment](https://github.com/nimiq/wallet/pull/72/commits/02906e011bb07416d0fbdff3f63ed0aa5b8731d8#diff-a90bbf62b9d812fd389c58b3a09c10f35a5eb609278359146245cecbb7abf06cL147) should be put here, if it still applies.
(Kind of confusing that `'false'` will lead to the same behavior).

onmax replied:
You are right, adding a `'false'` as a string will have the same behaviour, although I've added a comment 😸

onmax committed:
added comment to explain why Location.params is a string

onmax commented:
Just realized that there are some useless props. I remove them in the last commit 3ed4a25
